### ISSUE TITLE
Galaxy-Inspired Features to Consider

### ISSUE THREAD

**Issue by jmchilton** [common-workflow-language/common-workflow-language#27]
Very preliminary list to consider (and unfortunately probably biased toward the kinds of things @jmchilton thinks about on the Galaxy project). In order of increasing of controversy:

 - DOIs for citing underlying application on the tool.
 - Annotating inputs and outputs with platform-specific type extensions (URL's were mentioned at 2015/03/17 meeting - this is fine but documentation or example would be ideal).
 - Version for the tool (not the underlying application).
 - Requirement list - combinations of abstract package name and version that can be used by platforms to configure the environment for a particular tool. (This would be if Docker is not available).

Some more ideas of things I would like for Galaxy users but might not make sense to put in the draft - platform specific extensions (#26) may be better for these and advice on how to do that would be most appreciated. 

 - BibTeX for citing underlying tool on application.
 - Way to build a command for fetching the version of the underlying application (e.g. ``application --version``).

Then platform specific UI hints are huge - 

There are existing issues for some other features - in particular
 - Annotating datatypes on inputs and outputs - #7
 - Conditional outputs - #14 (the purposed flexibility will be a real challenge to represent in Galaxy - but Galaxy's variant is tied to Python and Galaxy data structures so is not appropriate).

**Comment by ntijanic** [comment]
Re DOIs, since we'll likely want more metadata for the underlying app, perhaps we can use a nested DOAP object:

```yaml
class: CommandLineTool
label: "samtools sort"
application:
  id: doi:10.1000/1234
  doap:name: samtools
  doap:description: "Samtools is..."
  doap:license: spdx:MIT
  doap:homepage: "http://samtools.sourceforge.net/"
  doap:repository: "https://github.com/samtools/samtools"
  doap:revision: "1.2"
```

Re version of the tool document, I'm not sure how to approach this. I'd prefer if the top-level identifier was a URL for the specific (immutable) revision of the document and use some other property to signify that e.g. "samtools_sort_0.1" and "samtools_sort_0.2" are different revisions of "samtools_sort". Perhaps use top-level `doap:name` and `doap:revision` for this purpose?

Re custom environment requirements, it should be done via custom value for the `class` property in the `requirements` list. For example:

```yaml
class: CommandLineTool
label: wget
requirements:
  - class: galaxy:AptPackageRequirement
    galaxy:package_name: wget
```

For the above tool, a platform that doesn't recognize `galaxy:AptPackageRequirement` should refuse to run it.

If you want to leave optional execution environment "hints", use the `hints` property:

```yaml
class: CommandLineTool
hints:
  - class: galaxy:PrefferedEC2InstanceType
    value: m1.large
```

For the above tool, a platform that doesn't recognize `galaxy:PrefferedEC2InstanceType` can ignore it.

There's currently no way to specify alternate sets of requirements, but it feels like it's something we can add later while keeping backward-compatibility (e.g. though a OneOfRequirement type).

Regarding conditional outputs, personally, I don't think we should use them in the manner proposed in that issue. Rather, they could simply be UI hints.

**Comment by mr-c** [comment]
From this week's call I think we identified a general need to specify how to extract a tool's version; should this be split out into its own issue?

**Comment by ntijanic** [comment]
Yeah, we should start splitting feature requests/ideas in own issues.

**Comment by smoe** [comment]
Somehow I feel that there are some things mixed up. A CWL command line describes just that. A command line may require a particular software to be installed or that executed tool already comes with the base installation on a particular platform. For instance above referenced wget does not need to get installed on Mac OS X, if I am not erroneous.  The workflow shoud just work, then, and not be ignored.

The challenge is to clarify that the wget the workflow demands is the same that a distribution's cognate package provides. This is not always the case. For excample it is sheer luck that I was once the first to provide a package called muscle, which is also a communication thing. And we had a conflict with plink which is also in the package putty. All these subtle differences across platforms shall not be expressed in the command line description itself. Instead, the command should refer to an abstract namespace/registry to identify the means to install a software that will provide the right binary.

Different platforms then make different proposals on how to install the command. And one platform could inherit such proposals from another. This will come particularly handy for cloud images that will differ in their wealth of packages that these already provide, but a particular resource may declare to be a derivative of Debian stable, such that anything not already provided would then be installed the same way that it would on any other Debian stable release.
